### ISSUE TITLE
Connection refused when running starting up Firefox

### ISSUE THREAD

**Issue by eweseong** [mozilla/geckodriver#979]
## System
* Version: 0.19.0
* Platform: Red Hat Enterprise Linux Server release 7.2 (Maipo)
* Firefox: 56
* Selenium: 3.6.0
* Python: 2.7.5
* Robotframework: 3.0.2
* robotframework-selenium2library: 3.0.0b1

## Testcase

I'm running some robot script which opens up a web page in Firefox, maximize the browser and closes the browser. On occasion the "Open Browser" will open up the Firefox without inserting the address provided and hangs there without returning error/timeout. The error happens intermittently and it is reproducible by running the following scripts.

[run batch_pybot.zip](https://github.com/mozilla/geckodriver/files/1343176/run.batch_pybot.zip)

## Trace-level log

[geckodriver.log](https://github.com/mozilla/geckodriver/files/1343171/geckodriver.log)

**Comment by whimboo** [comment]
I would like to see how your actual test looks like. I ask because in the log I do not see anything about the executed commands. I believe that a former instance of Firefox is not getting shutdown correctly and as such the next one cannot setup the server because the same port is used.

**Comment by eweseong** [comment]
I've attached the test case ran in the original comment under "run batch_pybot.zip"

**Comment by whimboo** [comment]
So I have never seen this framework, and as such it doesn't give me the necessary information to act on the question. Please really provide a trace level log from such a run. See the README on this repository in how to do it.

**Comment by eweseong** [comment]
I believe the log attached is a trace level log, I've followed the instruction and passed the argument -vv when starting up geckodriver process. Could you advice if any other step is needed? The test case essentially is using selenium to open a browser and maximize that browser, nothing more nothing less

**Comment by andreastt** [comment]
Judging by the attached log it just looks like Marionette isn’t able to bind to the given port:

```
JavaScript error: chrome://marionette/content/server.js, line 337: NS_ERROR_SOCKET_ADDRESS_IN_USE: Component returned failure code: 0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE) [nsIServerSocket.initSpecialConnection]
```

This could indicate that you have something already running on the port Firefox attempts to bind to. You said this happened only occasionally? That could be an indication of https://bugzilla.mozilla.org/show_bug.cgi?id=1240830.

**Comment by whimboo** [comment]
Also the batch file as contained in the zip archive contains a loop. When does the above issue appear? Already for the first iteration, or later on?

**Comment by eweseong** [comment]
> Also the batch file as contained in the zip archive contains a loop. When does the above issue appear? Already for the first iteration, or later on?
It is intermittent so it may happen on any iteration. From my result, it manage to execute 200 times before hitting the error.

**Comment by whimboo** [comment]
In that case it would be great to also see a trace log from the previous iteration. It could give an idea what was wrong in shutting down the browser.

**Comment by eweseong** [comment]
The attached log has the 2 previous run log before hitting the error
Line 1:T-2
Line 308: T-1
Line 612: Actual error run
[geckodriver.log](https://github.com/mozilla/geckodriver/files/1358079/geckodriver.log)

**Comment by whimboo** [comment]
Ok, so there is one difference in the log:

Here run - 2:
```
1507172723118	Marionette	TRACE	0 <- [1,6,null,{"cause":"shutdown"}]
1507172723159	geckodriver::marionette	TRACE	<- [1,6,null,{"cause":"shutdown"}]
1507172723159	webdriver::server	DEBUG	Deleting session
1507172723159	geckodriver::marionette	DEBUG	Stopping browser process
1507172723198	webdriver::server	DEBUG	<- 200 OK {"value": {}}
```

The shutdown is all fine and the process quit normally.

Here run - 1:
```
1507172730653	Marionette	TRACE	0 <- [1,6,null,{"cause":"shutdown"}]
1507172730699	geckodriver::marionette	TRACE	<- [1,6,null,{"cause":"shutdown"}]
1507172730699	webdriver::server	DEBUG	Deleting session
1507172730699	geckodriver::marionette	DEBUG	Stopping browser process
[Child 23731] WARNING: pipe error (3): Connection reset by peer: file /home/worker/workspace/build/src/ipc/chromium/src/chrome/common/ipc_channel_posix.cc, line 353
1507172730747	webdriver::server	DEBUG	<- 200 OK {"value": {}}
```

We triggered a shutdown of Firefox, but immediatelly seem to have killed the browser given that there is a IPC disconnect. That should usually not happen. I filed [bug 1403923](https://bugzilla.mozilla.org/show_bug.cgi?id=1403923) a week ago which seem to be the case you are stumbling in here.

Then in the failing run we get:
```
JavaScript error: chrome://marionette/content/server.js, line 337: NS_ERROR_SOCKET_ADDRESS_IN_USE: Component returned failure code: 0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE) [nsIServerSocket.initSpecialConnection]
JavaScript error: chrome://marionette/content/server.js, line 337: NS_ERROR_SOCKET_ADDRESS_IN_USE: Component returned failure code: 0x804b0036 (NS_ERROR_SOCKET_ADDRESS_IN_USE) [nsIServerSocket.initSpecialConnection]
1507172737009	geckodriver::marionette	TRACE	  connection attem
```

The second listed error is due to [bug 1254136](https://bugzilla.mozilla.org/show_bug.cgi?id=1254136) which should be fixed for Firefox 56.0. But why we get this error at all means that there must be a remaining (zombie) process of Firefox around.

@eweseong would you mind to enhance your shell script to print out all Firefox processes first before starting the next iteration? If there is a remaining process please check how long it takes for Firefox to quit, or if it does at all.

**Comment by whimboo** [comment]
I think that this is a dupe of issue #285.

**Comment by eweseong** [comment]
@whimboo I've give it a run and there's no active Firefox process before the error iteration run

**Comment by whimboo** [comment]
At line 307 I can already see such an IPC error. Sadly you are not printing out the ps output there. Mind updating your script so it prints it for each iteration at the beginning and end of the steps? I would kinda like to see the state of the child process. Maybe you can also include the geckodriver process, and a netcat output for port 2828.

But in general it confirms my suspicion from earlier today.

**Comment by whimboo** [comment]
Btw. can you also please test with Firefox 56, and check if the problem can still be seen?

**Comment by eweseong** [comment]
@whimboo The attached log is ran with Firefox 56, however with FF 56 the geckodriver process now hangs without attempting the full 600 atempts
[geckodriver.log](https://github.com/mozilla/geckodriver/files/1361366/geckodriver.log)

**Comment by whimboo** [comment]
@eweseong, thank you for the update. It's good to know that there is really no other Firefox process blocking this port. But that might because the test was run with Firefox 56 now.

By inspecting the last log it strongly reminds me on [bug 1397201](https://bugzilla.mozilla.org/show_bug.cgi?id=1397201). But that is Windows and not Linux.

Would it be possible for you to maybe trace the socket connection via Wireshark or another networking tool? I would really be interested if there is something blocking the Marionette server to receive commands. 

Something I totally missed earlier on this issue is that the port gets dynamically assigned. So only checking for port `2828` won't help at all. Sorry. Important would be the port of the former iteration as logged by Marionette via: `Marionette	INFO	Listening on port 46860`.

**Comment by gbajson** [comment]
@whimboo
I am able to reproduce this problem. 
I saved saved traffic into pcap file (attached).
[firefox.pcap.gz](https://github.com/mozilla/geckodriver/files/1401066/firefox.pcap.gz)

**Comment by whimboo** [comment]
@gbajson I don't see any failures in the attached trace level log. So I doubt that you are hitting the same issue as @eweseong. If there are problems for you please file a separate issue. Thanks.

**Comment by eweseong** [comment]
@whimboo I've noticed the [bug 1397201](https://bugzilla.mozilla.org/show_bug.cgi?id=1397201) has been fixed but i did not see the target version on it, would it included in the next release?

**Comment by whimboo** [comment]
@eweseong, this bug is not really related to your issue. It only draws some lines. In any case I would be interested in a new trace log while running Firefox 58 beta. Could you please do that for me?

**Comment by eweseong** [comment]
Sorry for the late reply, attached is the latest log run with Firefox 58 beta. The error attempt can be found starting @ line 602.
[n-3-geckodriver.log](https://github.com/mozilla/geckodriver/files/1516677/n-3-geckodriver.log)

as for your previous comment 
> trace the socket connection via Wireshark or another networking tool?

is there any particular protocol/port range you are trying to track?

**Comment by whimboo** [comment]
Ok, so that log was indeed pretty helpful. Here the interesting lines:

> 1512011499064	Marionette	DEBUG	Received observer notification "profile-after-change"
> 1512011499132	Marionette	DEBUG	Received observer notification "command-line-startup"
> 1512011499132	Marionette	INFO	Enabled via --marionette
> 1512011499148	geckodriver::marionette	DEBUG	Connected to Marionette on localhost:38128
> [..]
> 1512011501427	Marionette	DEBUG	New connections are accepted
> 1512011501428	Marionette	INFO	Listening on port 38128

As you can see above geckodriver makes a connection to Marionette before it actually has finalized its initialization. So a connection is being made to port 38128 at 1512011499148, which is 2.2s before Marionette accepts connections.

So I would like to know to which process geckodriver connects to. In your log there is no output from `ps` for any Firefox process. But given that geckodriver connects to something, there must be one.

We also see this very rarely in our automation ([bug 1400819](https://bugzilla.mozilla.org/show_bug.cgi?id=1400819#c9) but there was no way for me yet to isolate that. So if you get into such a situation again I really appreciate your feedback.

**Comment by whimboo** [comment]
Oh, and what is strange, each of your iterations use a different port number. So why is geckodriver connecting to a port which hasn't been used before?

If doable please try to run the following debug version of Firefox Nightly, which will spit out way more details to the log. Maybe there is something helpful for us included:

https://queue.taskcluster.net/v1/task/cbY2rrQXRC6QD-O1fM7zLQ/runs/0/artifacts/public/build/target.tar.bz2

Thanks!

**Comment by eweseong** [comment]
> Oh, and what is strange, each of your iterations use a different port number. So why is geckodriver connecting to a port which hasn't been used before?

hmm not too sure, I've not explicitly assign any ports so I would assume it is assigned by the driver randomly

Attached is the log ran with the nightly provided, the error attempt can be found starting @ line 1074
[n-3-geckodriver.log](https://github.com/mozilla/geckodriver/files/1530070/n-3-geckodriver.log)

**Comment by whimboo** [comment]
Ok, so Firefox experiencing a crash during shutdown, maybe related to the kill signal as sent to the process. But compared to other iterations this is interesting:

> Sleeping for 300 seconds.
> Type 'gdb /usr/lib64/firefox/firefox 400' to attach your debugger to this thread.

The process has been put in sleep mode, but your script is continuing. Additionally I can see dozen of different content processes of Firefox lingering around. That is something your previous logs haven't shown. Can you please check the full log, specifically when the first content processes have not been closed? Does it make a difference when you add a sleep (maybe 10s) between each iteration? Or do those older content processes live forever? This sounds heavily like a nasty bug in Firefox.

**Comment by eweseong** [comment]
> specifically when the first content processes have not been closed?

after the initial run

> do those older content processes live forever?

fortunately they will be killed after a certain amount of time ~5 - 10 minutes

> Does it make a difference when you add a sleep (maybe 10s) between each iteration?

i'll give this a try

**Comment by whimboo** [comment]
Hm, maybe you are affected by [bug 1404190](https://bugzilla.mozilla.org/show_bug.cgi?id=1404190). Given that geckodriver currently doesn't correctly close Firefox but somewhat kills it on quit, it might run into this situation. I'm currently working on [bug 1403923](https://bugzilla.mozilla.org/show_bug.cgi?id=1403923) to fix geckodriver. Maybe we wait until this bug is fixed and then try again with a nightly build of geckodriver.

**Comment by andreastt** [comment]
@eweseong Can you try again with a fresh build of Nightly?

**Comment by whimboo** [comment]
Nope, that was a failure of mine. A nightly build of Firefox will not work here. He needs a new build of geckodriver.  So please download this file, and under `bin/` you will find the fixed version of geckodriver.

https://queue.taskcluster.net/v1/task/D8DVmtUSRmCODgrwlky5LQ/runs/0/artifacts/public/build/target.common.tests.zip

Please use that for testing this problem.

**Comment by eweseong** [comment]
> A nightly build of Firefox will not work here

Is there a specific firefox version i should test the fixes with?

**Comment by eweseong** [comment]
> https://queue.taskcluster.net/v1/task/D8DVmtUSRmCODgrwlky5LQ/runs/0/artifacts/public/build/target.common.tests.zip

A virus was detected in the file above

**Comment by whimboo** [comment]
No, all versions of Firefox will work with this fix in place. So whether you wait for the next geckodriver release, which should be out hopefully soon, or build it yourself.

Do you have more details about which type of virus (I strongly believe it is a false positive) is detected here by which av software? Info on that would be very welcome.

**Comment by eweseong** [comment]
>Do you have more details about which type of virus (I strongly believe it is a false positive) is detected here by which av software? Info on that would be very welcome.

https://www.f-secure.com/v-descs/other_w32_generic.shtml
I agree that it is most likely a false positive but nevertheless it has trigger the alert of our IT guys

**Comment by andreastt** [comment]
We’ve seen this with the 32-bit Windows binaries in the past, c.f. https://github.com/mozilla/geckodriver/issues/671 and https://github.com/mozilla/geckodriver/issues/481.

**Comment by eweseong** [comment]
> https://queue.taskcluster.net/v1/task/D8DVmtUSRmCODgrwlky5LQ/runs/0/artifacts/public/build/target.common.tests.zip

It seems that the fix did not resolve the problem, attached is the log file generated.
[geckodriver.log](https://github.com/mozilla/geckodriver/files/1645747/geckodriver.log)

**Comment by whimboo** [comment]
The fix solved the problem very well in geckodriver, but what you see now is a crash in Firefox. So which version of Firefox did you use here? Please paste the `Build from` link on the `about:buildconfig` page. I assume it is a debug build? Also did you ever run with a Firefox version, which wasn't affected by this problem? Maybe test old Firefox versions with the updated geckodriver binary.

If it is not a regression it would also be good if you could run a latest officially nightly build and give me the trace log too. Same for the debug build if not done yet.

From spotting at the crash I assume it is sandbox related. So you may try using the preference `security.sandbox.content.level` and set it to different levels (at the moment it should be from 1 to 5).

Thanks!

**Comment by eweseong** [comment]
> Build from link on the about:buildconfig page.

https://hg.mozilla.org/mozilla-central/rev/38f49346a200cc25492236c7b3c536fc835fe031

> you could run a latest officially nightly build 

It is the latest officially nightly build

> So you may try using the preference security.sandbox.content.level and set it to different levels (at the moment it should be from 1 to 5).

I'll give this a try

**Comment by whimboo** [comment]
Interesting that we get that less stack info for a Nightly build. Anyway, it would be great if you could test this debug build instead:

https://queue.taskcluster.net/v1/task/fQ3U-Y2tQsCXHRx1hopD6w/runs/0/artifacts/public/build/target.tar.bz2

It should give us the information we want. Also you might want to test at least with the 55 release of Firefox. Note, that formerly it was not working due to geckodriver, which is fixed with the binary you use now.

**Comment by eweseong** [comment]
> test at least with the 55 release of Firefox

[geckodriver_ff_v55.log](https://github.com/mozilla/geckodriver/files/1654876/geckodriver_ff_v55.log)

> you could test this debug build instead

[geckodriver_ff_v59_debug.log](https://github.com/mozilla/geckodriver/files/1654878/geckodriver_ff_v59_debug.log)

**Comment by whimboo** [comment]
@eweseong so Firefox 55 is not showing this problem? I cannot find a crash in that trace log. Or is there anything which is still not working?

**Comment by eweseong** [comment]
@whimboo yes, the same problem persisted in Firefox 55, a browser can be seen opened without any URL and the scripts hangs for waiting on the "Open Browser" keyword from robotframework

**Comment by whimboo** [comment]
Ok, so that seems to be something else which should have been fixed in later versions. Lets concentrate on Nightly for now. When the crash happens and it waits those 300s, can you please check if the created Firefox profile from geckodriver is still around? If yes, have a look into the `minidumps` subfolder inside that profile, and copy all the dmp files to a safe place, then package those and send the archive my way. Then I can check what this crash is about. Thanks.

**Comment by eweseong** [comment]
The `minidumps` folder seems to be empty at the point of crash, is there another alternative to collect the crash?

**Comment by whimboo** [comment]
Sadly not, because as it looks like we crash too early during startup of Firefox that the crash reporter has not been loaded yet.

@luser, do you have an idea how to get more details for this particular startup crash of Firefox 59? I tried to use the fix stack script from `python/rb` but that wasn't working for me.

**Comment by eweseong** [comment]
- Version: 0.20.0
- Platform: Red Hat Enterprise Linux Server release 7.2 (Maipo)
- Firefox: Mozilla Firefox 59.0a1
- Selenium: 3.6.0
- Python: 2.7.5
- Robotframework: 3.0.2
- robotframework-selenium2library: 3.0.0

attached is the log ran with the latest geckodriver v0.20 and firefox nightly
[geckodriver_v020_ff_59.log](https://github.com/mozilla/geckodriver/files/1827998/geckodriver_v020_ff_59.log)

**Comment by whimboo** [comment]
Ok, so there is not really a crash and it was a side-effect of killing the browser with the former geckodriver versions. But there is actually a suspicious line in that log:

> [Parent 19607, Main Thread] WARNING: failed to bind socket: file /builds/worker/workspace/build/src/netwerk/base/nsServerSocket.cpp, line 361

Something prevents Marionette from binding to the requested port. Please note that you are using a very outdated Nightly version (59.0a1). @eweseong please check again after upgrading.

Also for trace logs please only attach the part which shows the `quit` message of the last working test, and then the failing test. Thanks.

**Comment by eweseong** [comment]
- Firefox: Mozilla Firefox 61.0a1

[geckodriver.log](https://github.com/mozilla/geckodriver/files/1836180/geckodriver.log)

**Comment by eweseong** [comment]
- Firefox: Mozilla Firefox 61.0a1

[geckodriver.log](https://github.com/mozilla/geckodriver/files/1836275/geckodriver.log)

**Comment by eweseong** [comment]
Besides the problem in the above log file attached, I've notice occasionally the browser just open up without any action and hangs there indefinitely. From the log file of that run that hangs, I notice that the Geckodriver connected to the assigned port before Marionette started listening on the same exact port. Could this be a race condition that causes the indefinite hang? Attached is the log file for that browser hang issue mentioned.

[geckodriver_browser_hang.log](https://github.com/mozilla/geckodriver/files/1846084/geckodriver_browser_hang.log)

**Comment by whimboo** [comment]
The first log shows that Marionette fails to create the server socket, and as result directly forces a shutdown of Firefox. Because of that connection attempts fail and geckodriver finally fails with `Process unexpectedly closed with status 0`. It's kinda sad that we cannot trigger a shutdown of Firefox with a different exit code yet. But anyway, line 25 shows that the browser shutdown itself. So I have no idea why the server socket could not be created - btw. did you remove the output in-between those tests for active sockets? I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1449612 to add better logging output in case of a socket creation failure.

The second log is interesting but doesn't help that much at the moment. Please try again to output any running Firefox process, and used local ports before the next geckodriver instance starts Firefox.

**Comment by andreastt** [comment]
I think that’s a good point: an exit code would be the ideal IPC mechanism for communicating to geckodriver that Firefox exited because it was unable to bind to a port. I filed [bug 1449637](https://bugzil.la/1449637) about providing an XPCOM API for setting the Firefox exit code.

For as long as [bug 1449612](https://bugzil.la/1449612) ensures we log a good message to the correct level (`FATAL`) this would be an acceptable workaround.

**Comment by whimboo** [comment]
@eweseong when you find the time please re-test in latest Nightly which contains my fix for [bug 1449612](https://bugzilla.mozilla.org/show_bug.cgi?id=1449612). I would like to see if there is a real port conflict. If there is please make sure to keep the output from the Firefox process in the log which used that port. Thanks.

**Comment by gassho** [comment]
Hi,

I had the same random problem (around 1 time over 4).
When creating a new FirefoxDriver (new FirefoxDriver(firefoxOptions)), the browser opens, but the code hangs indefinitely.

When I added a firefox log file to debug the problem, the problem disappeared.
Finally, just adding following line resolved my problem:
`    System.setProperty("webdriver.firefox.logfile", "/dev/null");`

My tested environments:
    Firefox: 55 to 59.0.2
    Geckodriver: v0.20
    Windows 7 32 bits
    Selenium: 3.11.0
    JDK 8

Hope this can help.
Gassho

**Comment by whimboo** [comment]
Pretty interesting log. So in case of the hang Marionette connects to `localhost:44054` about 2s before the started Firefox instance is ready and the Marionette server component gets started:

> 1523340498672	geckodriver::marionette	DEBUG	Connected to Marionette on localhost:44054
> [..]
> 1523340499660	Marionette	DEBUG	Received observer notification sessionstore-windows-restored
> [..]
> 1523340500206	Marionette	INFO	Listening on port 44054

The port as used in the last successful iteration was `43531` so it's clearly different from the one causing a hang in the last run.

@eweseong one thing you should try is to not only print out the Firefox and geckodriver processes, but also a list of ports as used. I would like to know if there is the port listed, which causes an early connect when Firefox gets started. If there is absolutely nothing using that port, it has to be the current Firefox binary.

Would you also mind to reduce your testcase to maybe only create the driver and quit? When I have such a test locally I'm not able to get it to hang. Not sure if this might be a Linux only issue, and I'm on MacOS.

**Comment by eweseong** [comment]
@whimboo I've strip the test case to only create the web driver and quit; I've also printed the ports in used before the start of the execution. Hope that this helps.

[20180413-geckodriver.log](https://github.com/mozilla/geckodriver/files/1905817/20180413-geckodriver.log)

**Comment by whimboo** [comment]
So the port Marionette binds to (in this case 35380) is not in use before. Due to that I really, really wonder what makes geckodriver to break out of this loop before the wanted Marionette connection is ready: https://dxr.mozilla.org/mozilla-central/rev/325ef357e5b73d63794e47c02c7f8e7cf58ccb48/testing/geckodriver/src/marionette.rs#1346-1383

@eweseong could you perhaps create a network trace log? Would it contain entries which lists the ports as part of a specific process? I haven't done that myself yet.

Also maybe you can describe your system setup for those tests in detail? Maybe I missed something when I tried to reproduce it locally myself. Thanks.

@andreastt, do you have an idea?

**Comment by eweseong** [comment]
@whimboo by network trace log do you mean a wireshark capture?

My current system setup:
Version: 0.20.0
OS: Red Hat Enterprise Linux Server release 7.2 (Maipo)
OS type: 64 bit
Processor: Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz (4 core)
Firefox: Mozilla Firefox 61.0a1
Selenium: 3.8.1
Python: 2.7.5
Robotframework: 3.0.2
robotframework-selenium2library: 3.0.0
robotframework-seleniumlibrary: 3.0.1

**Comment by whimboo** [comment]
Yes, or something which simply logs any local port usage during the tests. You could then truncate and anonymize the data. I would like to see what happens with during the failure start with geckodriver and Firefox.

Also please note that your Python 2.7.5 install is totally outdated (May 15, 2013) and you should immediate upgrade it to Python 2.7.14 or similar.

**Comment by eweseong** [comment]
Version: geckodriver 0.20.1
Platform: CentOS Linux release 7.4.1708 (Core)
Firefox: Mozilla Firefox 61.0a1
Selenium: 3.11.0
Python: 2.7.14 (running in virtualenv)

@whimboo I might have left out that I'm running on python virtualenv, not sure if this info helps. I've also updated the virtualenv python to 2.7.14 as requested.

Geckodriver log
[geckodriver.log](https://github.com/mozilla/geckodriver/files/1937675/geckodriver.log)

Network trace log
[network_trace_log.zip](https://github.com/mozilla/geckodriver/files/1937679/network_trace_log.zip)

**Comment by eweseong** [comment]
I've isolated the network trace log to capture and reset on each run, perhaps these logs can provide more accurate information

[logs.zip](https://github.com/mozilla/geckodriver/files/1941643/logs.zip)

**Comment by whimboo** [comment]
@eweseong when I check the latest log files I cannot find the ports of geckodriver in the networking trace log. So do those really match?

Otherwise I wonder if you could just try to run your tests by hard-coding the Marionette port to 2828 as described here: https://dxr.mozilla.org/mozilla-central/source/testing/geckodriver/doc/CrashReports.md

I wonder if that helps us a bit in finding the underlying issue by getting rid of those random port assignments.

**Comment by eweseong** [comment]
> Otherwise I wonder if you could just try to run your tests by hard-coding the Marionette port to 2828 as described here: https://dxr.mozilla.org/mozilla-central/source/testing/geckodriver/doc/CrashReports.md

After following the suggestion mentioned here, I'm unable to reproducing the error even after 2500 tries. I've attached the virtualenv and script that was previously used to reproduced the error. Inside the zip file there's the virtualenv.tar and script.tar.
Steps
1. untar both tar to desired location
2. make sure the env have tshark/wireshark installed
3. if tshark/wireshark is not installed, remove the tshark section in the batch_python.sh
4. source the virtualenv at ../virtualenv/bin/activate
5. if you are running on vnc, the default display port is set to :1
6. execute the batch_python.sh

the poc.py is the script having the issue, fixed_port.sh is the recommendation to hard code the marionette port

[script_and_virtualenv.zip](https://github.com/mozilla/geckodriver/files/2007992/script_and_virtualenv.zip)

**Comment by dmitryvk** [comment]
Apparently, I've run into this issue. I've been able to get some logs that show interesting properties (but not full trace logs yet).
Basically, I've been instantiating Selenium FirefoxDriver in a loop and got to a hang.
Here's a gist with logs: https://gist.github.com/dmitryvk/e8665b2d3114c2d9ee37e78060e66f80
Interesting parts are:
from geckodriver:
```
1535632251416   Marionette      FATAL   Remote protocol server failed to start: Error: Could not bind to port 41340 (NS_ERROR_SOCKET_ADDRESS_IN_USE) (chrome://marionette/content/server.js:87:17) JS Stack trace: set acceptConnections@server.js:87:17
```
from netstat:
```
tcp 0 0 127.0.0.1:41340 127.0.0.1:41340 ESTABLISHED 21201/geckodriver
```

It seems that;
1) geckodriver somehow does not close its 'probing' socket
2) therefore marionette can not bind to this port
3) geckodriver connects to itself and hangs indefinitely

Can there be a failure or a race to close a socket?

**Comment by whimboo** [comment]
That would be very suspicious, but well nothing is un-impossible! @dmitryvk could you extract a log from Wireshark? Maybe this gives us more information. At least this is a very interesting point, I never thought about yet. Thanks a lot! @andreastt, please check the last comment. If we don't force close the socket, we should do it.

**Comment by dmitryvk** [comment]
I'll try to get more logs.

**Comment by dmitryvk** [comment]
Also might be related: `std::net::TcpListener::bind` not only binds the socket, it also starts listening on this port.

**Comment by whimboo** [comment]
So I tried to reproduce it in some way with the following example, but it shows that the first socket connection gets successfully closed:

https://play.rust-lang.org/?gist=bfc27e0a5bac33bf2ba522c4194d92f1&version=stable&mode=debug&edition=2015

@dmitryvk are you running the geckodriver processes in parallel too, or is the test execution all sequential? If it's in parallel I could imagine what it is.

**Comment by whimboo** [comment]
So maybe a fix which landed recently in geckodriver but hasn't been released yet, could have fixed it by hardening the handshake between geckodriver and Marionette. Please have a look at https://bugzilla.mozilla.org/show_bug.cgi?id=1461608#c2.

If you could tell me which platform you are on, I could give you a link to a geckodriver binary, which you could test.

**Comment by whimboo** [comment]
Maybe the patch on the above mentioned bug will give us a clue of what's going on.

**Comment by whimboo** [comment]
Please note that we released geckodriver 0.22.0 which has this mentioned fix included. Would you both mind to upgrade and test again? I'm really interested if that new version makes a difference for you.

**Comment by whimboo** [comment]
Sadly we haven't gotten a reply anymore on this issue. As such I'm going to close it. Note that you still have about a week to reply before it is getting locked. I'm happy to reopen if it is still happening.
